parsec/core/fs/workspacefs/workspacefile.py

```diff
@@ -0,0 +1,179 @@
+# Parsec Cloud (https://parsec.cloud) Copyright (c) AGPLv3 2019 Scille SAS
+
+import os
+import re
+from parsec.core.fs.exceptions import FSUnsupportedOperation, FSOSError
+
+
+class WorkspaceFile:
+    """methods of this object should act like trio async file :
+    "https://github.com/python-trio/trio/blob/master/trio/_file_io.py"
+    using transactions"""
+
+    def __init__(self, fd, transactions, mode: str = "r"):
+        self._fd = fd
+        self._transactions = transactions
+        self._open = True
+        self._offset = 0
+        self._firstWrite = True
+        self._fileLenght = -1
+        self._name = None
+        mode.lower()
+        # preventing to open in write and read in same time or write and append or open with no mode
+        if (("w" in mode or "a" in mode) and "r" in mode) or (
+            ("w" in mode and "a" in mode) or re.search("[arw]", mode) is None
+        ):
+            raise ValueError("must have exactly one of create/read/write/append mode")
+        # preventing to open with non-existant mode
+        elif re.search("[^arwb]", mode) is not None:
+            raise ValueError("invalid mode: ", mode)
+        self._mode = mode
+
+    async def __aenter__(self):
+        return self
+
+    async def __aexit__(self, *args):
+        self._open = False
+        await self.close()
+
+    async def __anext__(self):
+        return await self.readline()
+
+    async def close(self):
+        await self._transactions.fd_close(self._fd)
+
+    def closed(self):
+        return self._open is False
+
+    async def file_stats(self):
+        stats = await self._transactions.fd_info(self._fd)
+        return stats
+
+    def mode(self):
+        return self._mode
+
+    def name(self):
+        raise NotImplementedError
+
+    async def read(self, size: int = -1):
+        """
+        Raises:
+            FSUnsupportedOperation
+        """
+        # check if readable
+        if self.readable() is False:
+            raise FSUnsupportedOperation
+        currentOffset = self._offset
```

Comment:
`CurrentOffset` is a tmp var to store the offset because I am modifying it before returning the transactions_read. I will rename it as `current_offset`.

Reply:
`       current_offset = self._offset
        self._offset += await self._transactions.fd_write(self._fd, data, self._offset)
        nbWrittedBytes = self._offset - current_offset
        return nbWrittedBytes`
In write_bytes function I use it to calculate the number of bytes written by the fd_write transaction.
I will rename `nbWrittedBytes` as `nb_writted_bytes`.